Set the inverse record on find with multiple ids from has_many association

sekiyama58 opened the pull request:
### Summary

Previously, the `find` method of the `has_many` association set the inverse record to the `belong_to` association only when called with a single id, but it didn't set it with multiple ids:

```ruby
class User < ApplicationRecord
  has_many :posts
end

class Post < ApplicationRecord
  belongs_to :user
end

user = User.create!
posts = [Post.create!(user: user), Post.create!(user: user)]

results = user.posts.find(posts.pluck(:id))
results.map(&:user) #=> loads the user 2 times
```

This was because `except(:limit, :offset)` in `find_some_ordered` had created a new relation irrelevant to the original association relation.
This commit fixes the issue by allowing the `except` and `only` methods of the association relation to maintain the association.

Co-authored-by: Ryuta Kamizono <kamipo@gmail.com>

sekiyama58 committed:
Set the inverse record on find with multiple ids from has_many association

Previously, the `find` method on has_many association set the inverse
record to belong_to association only when called with a single id,
but it didn't set it with multiple ids:

```ruby
class User < ApplicationRecord
  has_many :posts
end

class Post < ApplicationRecord
  belongs_to :user
end

user = User.create!
posts = [Post.create!(user: user), Post.create!(user: user)]

results = user.posts.find(posts.pluck(:id))
results.map(&:user) #=> loads the user 2 times
```

This was because `except(:limit, :offset)` in `find_some_ordered` had
created a new relation irrelevant to the original association relation.
This commit fixes the issue by allowing the `except` and `only` methods
of the association relation to maintain the association.

Co-authored-by: Ryuta Kamizono <kamipo@gmail.com>

kamipo commented on activerecord/lib/active_record/relation/finder_methods.rb:
```diff
@@ -505,7 +505,9 @@ def find_some(ids)
-        result = except(:limit, :offset).where(primary_key => ids).records
+        # Keep the association relation when possible
+        relation = limit_value || offset_value ? except(:limit, :offset) : self
+        result = relation.where(primary_key => ids).records
```
Can you fix it in `relation_with` instead?

```diff
diff --git a/activerecord/lib/active_record/relation/spawn_methods.rb b/activerecord/lib/active_record/relation/spawn_methods.rb
index f44f3e9d01..8636b34434 100644
--- a/activerecord/lib/active_record/relation/spawn_methods.rb
+++ b/activerecord/lib/active_record/relation/spawn_methods.rb
@@ -69,8 +69,8 @@ def only(*onlies)
 
     private
       def relation_with(values)
-        result = Relation.create(klass, values: values)
-        result.extend(*extending_values) if extending_values.any?
+        result = spawn
+        result.instance_variable_set(:@values, values)
         result
       end
   end
```

kamipo commented on activerecord/test/cases/associations/inverse_associations_test.rb:
```diff
@@ -509,6 +509,24 @@ def test_find_on_child_instance_with_id_should_not_load_all_child_records
+    assert_predicate child.association(:human), :loaded?
+  end
+
+  def test_find_on_child_instance_with_id_should_set_inverse_instances
```
```suggestion
  def test_find_on_child_instance_with_ids_should_set_inverse_instances
```

sekiyama58 commented:
@kamipo  Thanks for your review!
I have updated the commit based on your suggestion and added Co-authored-by tag.
(Unfortunately some CI tests seem failing due to docker.io pull restriction..)

kamipo commented:
Thanks!
